commons/src/main/java/org/infinispan/commons/util/concurrent/BlockingAbortPolicy.java

```diff
@@ -0,0 +1,27 @@
+package org.infinispan.commons.util.concurrent;
+
+import java.util.concurrent.ThreadPoolExecutor;
+
+/**
+ * A handler for rejected tasks that runs the task if the current thread is a blocking thread otherwise it
+ * rejects the task.
+ * @author wburns
+ * @since 10.1
+ */
+public class BlockingAbortPolicy extends ThreadPoolExecutor.AbortPolicy {
+   private BlockingAbortPolicy() { }
+
+   private static final BlockingAbortPolicy INSTANCE = new BlockingAbortPolicy();
+
+   public static ThreadPoolExecutor.AbortPolicy getInstance() {
+      return INSTANCE;
+   }
+
+   @Override
+   public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+      if (executor.isShutdown() || Thread.currentThread().getName().startsWith("async")) {
+         super.rejectedExecution(r, executor);
```

Comment:
How about throwing `IllegalLifecycleStateException` instead of `RejectedExecutionException`?

Reply:
Sure

Reply:
Well actually, I should probably only throw that for it being shutdown.

Reply:
Yep :)